test/jdk/java/net/httpclient/HttpRequestNewBuilderTest.java

```diff
@@ -0,0 +1,272 @@
+/*
+* Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+* DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+*
+* This code is free software; you can redistribute it and/or modify it
+* under the terms of the GNU General Public License version 2 only, as
+* published by the Free Software Foundation.
+*
+* This code is distributed in the hope that it will be useful, but WITHOUT
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+* version 2 for more details (a copy is included in the LICENSE file that
+* accompanied this code).
+*
+* You should have received a copy of the GNU General Public License version
+* 2 along with this work; if not, write to the Free Software Foundation,
+* Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+*
+* Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+* or visit www.oracle.com if you need additional information or have any
+* questions.
+*/
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient.Version;
+import java.net.http.HttpHeaders;
+import java.net.http.HttpRequest;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import static java.net.http.HttpClient.Version.HTTP_2;
+import static java.net.http.HttpClient.Version.HTTP_1_1;
+
+import org.testng.annotations.Test;
+import org.testng.annotations.DataProvider;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertThrows;
+import static org.testng.Assert.assertTrue;
+
+/**
+* @test
+* @bug 8252304
+* @summary HttpRequest.NewBuilder(HttpRequest) API and behaviour checks
+* @compile --enable-preview -source ${jdk.version} HttpRequestNewBuilderTest.java
+* @run testng/othervm --enable-preview HttpRequestNewBuilderTest
+*/
+public class HttpRequestNewBuilderTest {
+   static final Class<NullPointerException> NPE = NullPointerException.class;
+   static final Class<IllegalArgumentException> IAE = IllegalArgumentException.class;
+
+   record NamedAssertion (String name, BiConsumer<HttpRequest,HttpRequest> test) { }
+   List<NamedAssertion> REQUEST_ASSERTIONS = List.of(
+           new NamedAssertion("uri",            (r1,r2) -> assertEquals(r1.uri(), r2.uri())),
+           new NamedAssertion("method",         (r1,r2) -> assertEquals(r1.method(), r2.method())),
+           new NamedAssertion("timeout",        (r1,r2) -> assertEquals(r1.timeout(), r2.timeout())),
+           new NamedAssertion("version",        (r1,r2) -> assertEquals(r1.version(), r2.version())),
+           new NamedAssertion("headers",        (r1,r2) -> assertEquals(r1.headers(), r2.headers())),
+           new NamedAssertion("expectContinue", (r1,r2) -> assertEquals(r1.expectContinue(), r2.expectContinue()))
+   );
```

Comment:
There should be an assertion for request bodies as well. You could check the following:

1. if r1 has a body, r2 must have a body
2. if r1 has no body, then r2 must not have a body
3. if both r1 and r2 have a body, then the body type & content length should be identical
4. you could subscribe to each body publisher and verify that the set of bytes you eventually get are identical

Reply:
Additional assertion added to check all 4 points raised see L109-142 in commit https://github.com/openjdk/jdk/pull/1059/commits/7928083ae954c44b5357064b86af0e1a3f53588b